Splice draft (#863)
lightningnetwork/lightning-rfc: 02-peer-protocol.md

  - MUST now consider the channel to be quiescing.
  - MUST NOT send an update message after `stfu`.

The receiver of `stfu`:
  - if it has sent `stfu` then:
    - MUST now consider the channel to be quiescent
  - otherwise:
    - SHOULD NOT send any more update messages.
    - MUST reply with `stfu` once it can do so.

Upon disconnection:
  - the channel is no longer considered quiescent.

### Rationale

The normal use would be to cease sending updates, then wait for all
the current updates to be acknowledged by both peers, then start
quiescence.  For some protocols, choosing the initiator matters,
so this flag is sent.

If both sides send `stfu` simultaneously, they will both set
`initiator` to `1`, in which case the "initiator" is arbitrarily
considered to be the channel funder (the sender of `open_channel`).
The quiescence effect is exactly the same as if one had replied to the
other.

## Splicing

Splicing is the term given for replacing the funding transaction with
a new one.  For simplicity, splicing takes place once a channel is
[quiescent](#channel-quiescence).

Operation returns to normal once negotiation is done (while waiting
for the splice transaction(s) to confirm).

> The splice is finally terminated when both sides send

Reviewer: There has been some discussion on this. What I believe makes most sense is disconnect is treated the same as `tx_abort` and results in dropping the splice up until the point generally when you have sent the signatures.

Some notes on the exact timing of when are here: https://github.com/lightning/bolts/pull/863/files#r897441047

   * [`channel_id`:`channel_id`]
   * [`u32`:`funding_feerate_perkw`]
   * [`point`:`funding_pubkey`]

1. type: 76 (`splice_ack`)
2. data:
   * [`chain_hash`:`chain_hash`]
   * [`channel_id`:`channel_id`]
   * [`point`:`funding_pubkey`]

#### Requirements

The sender of `splice`:
- MUST have successfully initiated quiescence.
- MUST NOT send `splice` before sending and receiving `funding_locked`.
- MUST NOT send `splice` while a splice is being negotiated.
- If any splice is in progress:
  - MUST NOT send a splice message with `funding_feerate_perkw` which is less than 1.25 the previous `funding_feerate_perkw` (rounded down).

The receiver of `splice`:
- SHOULD fail the connection if there is an ongoing splice, and the `funding_feerate_perkw` is not at least 1.25 the previous `funding_feerate_perkw` (rounded down).
- MUST respond with `splice_ack` containing its own `funding_pubkey`.
- MUST begin splice negotiation.

The receiver of `splice_ack`:
- MUST begin splice negotiation.

#### Rationale

There is no harm in agreeing to a splice with a high feerate
(presumably the recipient will not contribute to the splice which they
consider to be overpaying).

## Splice Negotiation

> The splice negotiation is very similar to the `init_rbf` negotiation:

Reviewer 2: https://github.com/lightning/bolts/pull/851/files#diff-ed04ca2c673fd6aabde69389511fa9ee60cb44d6b2ef6c88b549ffaa753d6afeR1087 for those following at home since this isn't defined in master or this branch

- MUST `tx_add_output` a zero-value output which pays to the two funding keys using the higher of the two `generation` fields.
- MUST pay for the common fields.

Upon receipt of consecutive `tx_complete`s, each node:
- MUST fail negotiation if there is not exactly one input spending the current funding transaction.
- MUST fail negotiation if there is not exactly one output with zero value paying to the two funding keys (a.k.a. the new channel funding output)
- MUST calculate the channel capacity for each side:
  - Start with the previous balance
  - Add that side's new inputs (excluding the one spending the current funding transaction)
  - Subtracting each sides new outputs (except the zero-value one paying to the funding keys)
  - Subtract the total fee that side is paying for the splice transaction.
- MUST replace the zero-value funding output amount with the total channel capacity.
- MUST calculate the channel balance for each side:
  - Subtract any outstanding HTLCs offered by that side.
- if either side has added an output other than the new channel funding output:
  - MUST fail the negotiation if the balance for that side is less than 1% of the total channel capacity.

> - SHOULD NOT fail if the splice transaction is nonstandard.

Reviewer 3: Why?

A counterparty could make a tx that cant be broadcast under policy rules and let my node revoke the previous commitment. My node would have 0 valid commitment txs then

> - MUST increment the commitment number and send `commitment_signed`, including the signatures for the splice transaction.

Reviewer: In order to prevent awkward states requiring force closures, double spending inputs, etc:
* The first sender of `commitment_signed` must be the same as the first sender of `tx_signatures`
* The second sending of `commitment_signed` commits that sender to be unable to `tx_abort`
* At this point both sides should have data committed to DB such that they can resume signing after a reconnect
* The first sender can now safely send `tx_signatures` or `tx_abort`
* After this point neither side can use `tx_abort`

This is needed to handle cases where `tx_signatures` is sent but not received. The sender is left in a vulnerable position where their peer may be able to broadcast the tx but they cannot. This leaves them forced to search the blockchain for the splice forever, have another successful splice, double spend an input, or force close the channel. Even if they find the splice onchain the recourse in that case would be to force close causing awkward competing transactions.

As such `tx_abort` after the signature is sent is a hostile act and shouldn't be allowed per the spec. Since `tx_signatures` can't be guaranteed to be received after being sent, the final `commitment_signed` message acts as an ack commiting the peer to signing.

I believe the correct response to receiving an invalid `tx_abort` is to:
1) Force close
2) Wait for force close to complete, while,
3) Also monitoring for the splice confirmation onchain, and,
4) Force closing that splice confirmation if it appears onchain

** Allowed examples **

```
commit-sig -> 
           <- commit-sig
tx-sig     ->
           <- tx-sig
```

```
commit-sig -> 
           <- commit-sig
tx-abort   ->
```

** Disallowed examples **

```
commit-sig -> 
           <- commit-sig
           <- tx-abort
```

```
commit-sig -> 
           <- commit-sig
tx-sig     ->
           <- tx-abort
```

```
commit-sig -> 
           <- commit-sig
tx-sig     ->
           <- tx-sig
tx-abort   ->
```

```
commit-sig -> 
           <- commit-sig
tx-sig     ->
tx-abort   ->
```

Reviewer 4: I don't understand why we need to increment the `commitment_number`, this requires exchanging a lot of data that is completely unnecessary:

- we need to re-sign every other active commitment, but they didn't change
- we need to exchange `revoke_and_ack` for the previous commitment, but there's no good reason for that since there was no change

It looks like we add this unnecessary complexity only to be able to re-use the existing `channel_reestablish` fields (`next_commitment_number` / `next_revocation_number`) to figure out that a splice's `commit_sig` needs to be retransmitted when reconnecting.

I think it's much simpler to send a single `commit_sig` with the current `commitment_number` for the splice transaction. The only drawback is that if we get disconnected, we need to use a different mechanism to figure out whether that `commit_sig` must be retransmitted. But that mechanism should be really simple, as only the following scenarios are possible:

- there is no ongoing splice -> nothing new to retransmit
- there is an ongoing splice: peers only need to tell their counterparty whether they received their splice `commit_sig` or not (since we're in `stfu` mode, there cannot be "normal" operations to retransmit, the commitment numbers will be up-to-date on both sides)

So it feels like the addition of a single boolean to `channel_reestablish` lets us trivially deal with the reestablish logic.

> The first sender of commitment_signed must be the same as the first sender of tx_signatures
> The second sending of commitment_signed commits that sender to be unable to tx_abort
> At this point both sides should have data committed to DB such that they can resume signing after a reconnect

As detailed [here](https://gist.github.com/t-bast/2594ecf0314f0b91293a3f5e3bf10540), I think peers MUST store the channel state right after _sending_ `commit_sig`, otherwise there are cases where one side has stored the channel and the other hasn't, which leads to an unnecessary force-close. Once peers do that, we don't care about the ordering of `commit_sig`, I don't think we need to add any requirement for that.

> - Upon receipt of `revoke_and_ack` for the previous commitment:

Reviewer 5: I know this is an old draft, but I think the `revoke_and_ack` should be the last step:

```
--cs----->
<-cs------
--tx_sig-->
<-tx_sig---
--rev----->
<-rev-----
```
Otherwise, it's possible that one side has revoked their prior commitment and now has a commit_sig for a tx they can't broadcast as they don't have tx_sig yet.

Reviewer: Thanks for pointing this out.

The current flow is:
```
--cs----->
<--rev---
<--cs----
---rev--->
```

Then later we do the `tx_sig` stuff.

The rev messages are implied here but are specified elsewhere in the spec. Here's the current process showing how the rev messages are handled: https://github.com/ddustin/lightning/blob/5b655039d0f87acf2b2fec2cac11b999ab2790f3/channeld/channeld.c#L2500

Could be worth adding the rev's to the spec here for additional clarity 🤷

Reviewer 3: B _gets_ a signature from A (and has 2 non-revocable commitments then), then B revokes the previous one. one of the parties either has 1 or 2 valid commitments, not 0

Reviewer 5: If I'm understanding this correctly, B's commit_sig that A sends is spending from the intermediate splice transaction. It is not usable until B has the tx_sig from A to broadcast the splice transaction.

Reviewer 3: oh you're right, the flow in your [first comment](https://github.com/lightning/bolts/pull/863#discussion_r910126737) is the way to go then, or to make it a bit faster:
```
--cs------>
<-cs-------    <-- 2 messages at once
<-tx_sig---
--tx_sig-->    <-- 2 messages at once
--rev----->
<-rev------
```
but this can make things more complicated

Reviewer 5: Think my comment is wrong here since the `commit_sig` actually includes a signature spending from the original funding outpoint and a signature (in the tlv) spending from the new funding outpoint

Reviewer 3: if that is the case i think this line has to be made clearer

```
- MUST increment the commitment number and send `commitment_signed`, including the signatures for the splice transaction.
```

are these "signatures for the splice transaction" actually signatures for the commitment transaction that spends the splice transaction?

Reviewer: They're incrementing the commitment tx signatures (not signing the splice tx itself -- there's no count for those that I'm aware of).

I believe the commitment number is shared across both original funding and splice funding.

Reviewer: > if that is the case i think this line has to be made clearer
> 
> ```
> - MUST increment the commitment number and send `commitment_signed`, including the signatures for the splice transaction.
> ```
> 
> are these "signatures for the splice transaction" actually signatures for the commitment transaction that spends the splice transaction?

These are the signatures for the commitment transaction child of the splice -- I'll update it to make that more clear 👍

  - the channel is no longer considered quiescent.

### Rationale

The normal use would be to cease sending updates, then wait for all
the current updates to be acknowledged by both peers, then start
quiescence.  For some protocols, choosing the initiator matters,
so this flag is sent.

If both sides send `stfu` simultaneously, they will both set
`initiator` to `1`, in which case the "initiator" is arbitrarily
considered to be the channel funder (the sender of `open_channel`).
The quiescence effect is exactly the same as if one had replied to the
other.

## Splicing

Splicing is the term given for replacing the funding transaction with
a new one.  For simplicity, splicing takes place once a channel is
[quiescent](#channel-quiescence).

Operation returns to normal once negotiation is done (while waiting
for the splice transaction(s) to confirm).

The splice is finally terminated when both sides send
`splice_complete` to indicate that one of the splice transactions
reached acceptable depth.

### The `splice` Message

1. type: 74 (`splice`)
2. data:
   * [`chain_hash`:`chain_hash`]
   * [`channel_id`:`channel_id`]
   * [`u64`:`funding_satoshis`]
   * [`u32`:`funding_feerate_perkw`]

>    * [`point`:`funding_pubkey`]

Reviewer: In the equivalent dual funding spec we pass locktime in the initiation `open_channel2` [spec](https://github.com/lightning/bolts/pull/851/files#diff-ed04ca2c673fd6aabde69389511fa9ee60cb44d6b2ef6c88b549ffaa753d6afeR1115).

In this spec we need to either require locktime to be set to a specific value (say 0) or allow the initiator to specify one in this message so the peer's commitments are guaranteed to match.

This conversation probably came up during the dual funding process. Any thoughts on the best approach here @niftynei @t-bast?

Reviewer 4: I think we should do the same for splice, the peer initiating the splice should choose the `lockTime`.

Reviewer: Okay sounds like the consensus is support locktime and let nodes choose their own policy of what locktimes to accept.

The quiescence effect is exactly the same as if one had replied to the
other.

## Splicing

Splicing is the term given for replacing the funding transaction with
a new one.  For simplicity, splicing takes place once a channel is
[quiescent](#channel-quiescence).

Operation returns to normal once negotiation is done (while waiting
for the splice transaction(s) to confirm).

The splice is finally terminated when both sides send
`splice_complete` to indicate that one of the splice transactions
reached acceptable depth.

### The `splice` Message

1. type: 74 (`splice`)
2. data:
   * [`chain_hash`:`chain_hash`]
   * [`channel_id`:`channel_id`]
   * [`u64`:`funding_satoshis`]
   * [`u32`:`funding_feerate_perkw`]
   * [`u32`:`locktime`]
   * [`point`:`funding_pubkey`]

1. type: 76 (`splice_ack`)
2. data:
   * [`chain_hash`:`chain_hash`]
   * [`channel_id`:`channel_id`]
   * [`u64`:`funding_satoshis`]
   * [`point`:`funding_pubkey`]

`funding_satoshis` is the amount the sender is putting into the
post-splice channel. It includes their old funded channel amount
plus any satoshis they intend to add, or, less any satoshis they

> intend to remove.

Reviewer: > What is meant by "old funded channel amount"? Is this the sender's local channel balance, their previous contribution to the funding transaction, or something else?

This is the balance before the splicing activity began.

> intend to remove.

Reviewer: This is getting reworked to be a relative satoshi changed to the channel instead of a total amount -- more discussion on the litdevs mailing list.

#### Requirements

The sender of `splice`:

> - MUST have successfully initiated quiescence.

Reviewer 6: to clarify: does that mean that only the initiator of `stfu` will be allowed to advance to send `splice`, or any other protocol for that matter?  (i.e. if Alice sends `stfu` to Bob, only Alice can send `splice`)

what happens in the event where Alice could send `stfu`, refuse to advance to the `splice` (or any other protocol) -- will they be taking the channel hostage?

Reviewer: Yes the initiator should be the one to start the splice. I like this approach because keeping turn order explicit keeps the protocol clean and reduces complexity.

Starting `stfu` and hanging is equivalent to hanging on a channel generally. Nodes should go through their normal logic for a hanging node (ie. reconnect after a timeout, eventually force close). `stfu` mode should be reset on reconnection.

Reviewer 7: > what happens in the event where Alice could send stfu, refuse to advance to the splice (or any other protocol) -- will they be taking the channel hostage?

As @ddustin mentions, your channel peer can choose to stop responding at any point (after receiving `add_htlc` for ex) and produce an equivalent result: the channel becomes useless and any in-flight payments are stuck.

`stfu` is no different, and it protocol timing failures (takes too long to advance etc) should be dealt with in the same manner that your channel currently handles other failures to respond.

- MUST NOT send `splice` before sending and receiving `funding_locked`.
- MUST NOT send `splice` while a splice is being negotiated.
- If any splice is in progress:
  - MUST NOT send a splice message with `funding_feerate_perkw` which is less than 1.25 the previous `funding_feerate_perkw` (rounded down).

The receiver of `splice`:
- SHOULD fail the connection if there is an ongoing splice, and the `funding_feerate_perkw` is not at least 1.25 the previous `funding_feerate_perkw` (rounded down).

> - MUST respond with `splice_ack` containing its own `funding_pubkey`.

Reviewer 7: Yes! I think this is correct. Generally when thinking about `splice_init` and `splice_ack`, I consider them to be analogs of `open_channel` and `accept_channel`.

I like the format that they use for rejecting/accepting `opens` fwiw https://github.com/lightning/bolts/blob/master/02-peer-protocol.md#requirements-1

- MUST begin splice negotiation.

The receiver of `splice_ack`:
- MUST begin splice negotiation.

#### Rationale

There is no harm in agreeing to a splice with a high feerate
(presumably the recipient will not contribute to the splice which they
consider to be overpaying).

## Splice Negotiation

The splice negotiation is very similar to the `init_rbf` negotiation:
both sides alternate sending `tx_add_input` and `tx_add_output` until
they both send consecutive `tx_complete`.

### Requirements

(Note that many of these messages have their own, additional
requirements detailed elsewhere).

The initiator:
- MUST `tx_add_input` an input which spends the current funding transaction output.

> - MUST `tx_add_output` a zero-value output which pays to the two funding keys using the higher of the two `generation` fields.

Reviewer: This appears to be referring to an old version of the splicing spec (https://github.com/lightning/bolts/commit/f3a800699a8cd54d425238bc8b685a1636a1012b).

Since we're using explicit pubkeys now this reference should probably be removed.

> - MUST `tx_add_output` a zero-value output which pays to the two funding keys using the higher of the two `generation` fields.

Reviewer 4: Yes I don't see why we'd need that step at all, that should be removed.

- MUST pay for the common fields.

Upon receipt of consecutive `tx_complete`s, each node:
- MUST fail negotiation if there is not exactly one input spending the current funding transaction.
- MUST fail negotiation if there is not exactly one output with zero value paying to the two funding keys (a.k.a. the new channel funding output)
- MUST calculate the channel capacity for each side:
  - Start with the previous balance
  - Add that side's new inputs (excluding the one spending the current funding transaction)
  - Subtracting each sides new outputs (except the zero-value one paying to the funding keys)
  - Subtract the total fee that side is paying for the splice transaction.

> - MUST replace the zero-value funding output amount with the total channel capacity.

Reviewer: Yeah great question. They're enforced later during signing and to a lesser extent during feerate checking.

It has to happen then because we can't know if they've satisfied the funding amount until both sides finish `tx_complete`.

> - MUST replace the zero-value funding output amount with the total channel capacity.

Reviewer 4: This should be removed as well.

  - MUST consider splice negotiation complete.
  - MUST consider the connection no longer quiescent.

On reconnection:
- MUST retransmit the last splice `tx_signatures` (if any).
- MUST ignore any redundant `tx_signatures` it receives.

### Rationale

If a side does not meet the reserve requirements, that's OK: but if
they take funds out of the channel, they must ensure that they do meet
them.  If your peer adds a massive amount to the channel, then you
only have to add more reserve if you want to contribute to the splice
(and you can use `tx_remove_output` and/or `tx_remove_input` part-way
through if this happens).

## Splice Completion

### The `splice_locked` Message

1. type: 77 (`splice_locked`)
2. data:
   * [`channel_id`:`channel_id`]

`channel_id` is the pre-splice channel's `channel_id`.

Each node:
- if any splice transaction reaches depth 6:
  - MUST send `splice_locked`.

Once a node has received and sent `splice_locked`:
  - Until sending OR receiving of `revoke_and_ack`
    - MUST ignore `announcement_signatures` messages where `short_channel_id`
      matches the pre-splice short channel id.
    - MUST ignore `commitment_signed` messages where `splice_channel_id`

>       does not match the `channel_id` of the confirmed splice.

Reviewer 8: Can we more explicitly define this `channel_id`? Is it just TXID based as usual? So even if we started with a dual-funded channel, where `channel_id` is XORed revocation basepoints, then after successful splice we now have a TXID-based one?

channel. To avoid this, the funder should ensure the funding transaction
confirms in the next 2016 blocks.

## Channel Quiescence

Various fundamental changes, in particular protocol upgrades, are
easiest on channels where both commitment transactions match, and no
pending updates are in flight.  We define a protocol to quiesce the
channel by indicating that "SomeThing Fundamental is Underway".

### `stfu`

1. type: 2 (`stfu`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u8`:`initiator`]

### Requirements

The sender of `stfu`:

>   - MUST NOT send `stfu` if any of the sender's htlc additions, htlc removals

Reviewer 9: I agree, the protocol specified it in the close channel, but to me sounds more natural have the requirements here 

See https://github.com/lightning/bolts/pull/863/files#r1334720458

    or fee updates are pending for either peer.
  - MUST NOT send `stfu` twice.
  - if it is replying to an `stfu`:
    - MUST set `initiator` to 0
  - otherwise:
    - MUST set `initiator` to 1

>   - MUST set `channel_id` to the id of the channel to quiesce.

Reviewer 9: If the funding transaction is not confirmed the channel is already disabled no? Maybe the `stfu` should not be sent in this case?

The receiver of `stfu`:
  - if it has sent `stfu` then:
    - MUST now consider the channel to be quiescent
  - otherwise:
    - SHOULD NOT send any more update messages.
    - MUST reply with `stfu` once it can do so.

Upon disconnection:
  - the channel is no longer considered quiescent.

### Rationale

The normal use would be to cease sending updates, then wait for all
the current updates to be acknowledged by both peers, then start
quiescence.  For some protocols, choosing the initiator matters,
so this flag is sent.

If both sides send `stfu` simultaneously, they will both set
`initiator` to `1`, in which case the "initiator" is arbitrarily
considered to be the channel funder (the sender of `open_channel`).
The quiescence effect is exactly the same as if one had replied to the
other.

## Splicing

Splicing is the term given for replacing the funding transaction with
a new one.  For simplicity, splicing takes place once a channel is
[quiescent](#channel-quiescence).

Operation returns to normal once negotiation is done (while waiting
for the splice transaction(s) to confirm).

The splice is finally terminated when both sides send

> `splice_complete` to indicate that one of the splice transactions

Reviewer 10: Is `splice_locked` meant here?

Reviewer: Ah yes great find

> reached acceptable depth.

Reviewer 11: I would strongly recommend to make `confirmation_depth` a parameter of `splice` to know ahead when the splicing is considered as “final” and the state can be forgotten.

The chain is almost the only reliable clock we have to sync lightning state machines.

Reviewer: Both sides need to see the splice be confirmed and exchange `splice_locked` before it can be forgotten

### The `splice` Message

> 1. type: 74 (`splice`)

Reviewer 10: Type `74` is used by `tx_abort` (interactive TX negotiation). What is the correct value here? 75? (as used by CLN)

Reviewer: This is a great question @t-bast what did you end up using to make this not conflict with `tx_abort`?

Reviewer 4: Our implementation doesn't use the spec message types to make it simpler to handle compatibility with the official version once we converge on one. This is somehow linked to using experimental feature bits (see my comment [here](https://github.com/lightning/bolts/pull/1108#issuecomment-1715870376) for details). We're currently using message type 37000 for `splice`.

I think we should use 80 and 81 for `splice` and `splice_ack` (this way we leave 75-79 for potential additions to `interactive-tx`). Or we use 76 and 77, whatever you think is best!

Reviewer: Using 80 and 81 sounds good to me! Nice round numbers 👍

2. data:
   * [`chain_hash`:`chain_hash`]
   * [`channel_id`:`channel_id`]
   * [`u64`:`funding_satoshis`]
   * [`u32`:`funding_feerate_perkw`]
   * [`u32`:`locktime`]
   * [`point`:`funding_pubkey`]

1. type: 76 (`splice_ack`)
2. data:
   * [`chain_hash`:`chain_hash`]
   * [`channel_id`:`channel_id`]
   * [`u64`:`funding_satoshis`]

>    * [`point`:`funding_pubkey`]

Reviewer 11: I think you will need a field to indicate the scriptpubkey / witnesscript / tree of tapscripts used as the funding scriptpubkey. I don’t think you wish to rule out the usage of P2WSH for now - if ever, especially if you have folks doing custom ECDSA protocols.

`funding_satoshis` is the amount the sender is putting into the
post-splice channel. It includes their old funded channel amount
plus any satoshis they intend to add, or, less any satoshis they
intend to remove.

#### Requirements

The sender of `splice`:
- MUST have successfully initiated quiescence.

> - MUST NOT send `splice` before sending and receiving `funding_locked`.

Reviewer 11: I think this should be update to `channel_ready` and explain if a splice of a non-confirmed channel is allowed or not.

- MUST NOT send `splice` while a splice is being negotiated.
- If any splice is in progress:
  - MUST NOT send a splice message with `funding_feerate_perkw` which is less than 1.25 the previous `funding_feerate_perkw` (rounded down).

The receiver of `splice`:
- SHOULD fail the connection if there is an ongoing splice, and the `funding_feerate_perkw` is not at least 1.25 the previous `funding_feerate_perkw` (rounded down).
- MUST respond with `splice_ack` containing its own `funding_pubkey`.
- MUST begin splice negotiation.

The receiver of `splice_ack`:
- MUST begin splice negotiation.

#### Rationale

There is no harm in agreeing to a splice with a high feerate
(presumably the recipient will not contribute to the splice which they
consider to be overpaying).

## Splice Negotiation

> The splice negotiation is very similar to the `init_rbf` negotiation:

Reviewer 12: ```suggestion
The splice negotiation is very similar to the `tx_init_rbf` negotiation:
```

Reviewer: Nice catch 👍

both sides alternate sending `tx_add_input` and `tx_add_output` until
they both send consecutive `tx_complete`.

### Requirements

(Note that many of these messages have their own, additional

> requirements detailed elsewhere).

Reviewer 9: This I think is required because the dual funding is still not merged, but we should define this and not point the reader around the spec.

Maybe I will also specify just the difference with the other definition(?)

- If recipient's sum(tx_add_input.amount) < peer's sum(tx_add_input.amount); or
if recipient's sum(tx_add_input.amount) == peer's sum(tx_add_input.amount) and
recipient is the `initiator` of the splice:
  - SHOULD send `tx_signatures` first for the splice transaction.
- else
  - MAY send `tx_signatures` first.

- Upon receipt of `tx_signatures` for the splice transaction:
  - MUST consider splice negotiation complete.
  - MUST consider the connection no longer quiescent.

On reconnection:
- MUST retransmit the last splice `tx_signatures` (if any).
- MUST ignore any redundant `tx_signatures` it receives.

### Rationale

If a side does not meet the reserve requirements, that's OK: but if
they take funds out of the channel, they must ensure that they do meet
them.  If your peer adds a massive amount to the channel, then you
only have to add more reserve if you want to contribute to the splice
(and you can use `tx_remove_output` and/or `tx_remove_input` part-way
through if this happens).

## Splice Completion

### The `splice_locked` Message

1. type: 77 (`splice_locked`)
2. data:
   * [`channel_id`:`channel_id`]

`channel_id` is the pre-splice channel's `channel_id`.

Each node:

> - if any splice transaction reaches depth 6:

Reviewer 11: I think really depth should be a `confirmation_depth` in the `splice` message. Some users might wish a lower or bigger amount, in function of the `funding_satoshis` committed in the channel and their liquidity prefs.

Reviewer: I believe as structured just sending `splice_locked` early (say 1 conf) should be fine. If the other node requires 6, it will remember you sent it and send it back after it sees 6.

It's important to resend it on restart as it's not a state that is preserved across restarts.

This behavior is already needed to handle blocks coming in at mismatched times.

Reviewer 13: we could make the depth of 6 a recommendation here

Reviewer 11: > I believe as structured just sending splice_locked early (say 1 conf) should be fine. If the other node requires 6, it will remember you sent it and send it back after it sees 6.

The `confirmation_depth` is a security parameter which should be scaled in function of the funds at stake. A miner can always reorg a low-depth confirmed splicing transaction to double-spend a high-value HTLC (and note the batching of block reorg against unbounded number of channels).

Of course, you can send the `splice_locked` once the depth has reached your local node satisfying value. However you’re depriving your counterparty of visibility on when the channel is ready to route / forward HTLC.

  - MAY send a `shutdown` before a `funding_locked`, i.e. before the funding transaction has reached `minimum_depth`.
  - if there are updates pending on the receiving node's commitment transaction:
    - MUST NOT send a `shutdown`.
  - if there is an ongoing splice:

>     - MUST NOT send a `shutdown`.

Reviewer 9: this can be moved as required of quiescence? Splice required that the channel is in quiet mode so this propriety will be already satisfied

Sounds more natural to me

Reviewer: Yes I agree

Reviewer: There may be a case where we still want something more sophisticated around `shutdown`, `stfu`, and `splice` down the road here. Long term the goal is to make channel closes also be splices so multiple can be merged together as well as closing directly into a channel open.

More discussion is needed to solve the complexities around this 🤔

Reviewer: > this can be moved as required of quiescence? Splice required that the channel is in quiet mode so this propriety will be already satisfied
> 
> Sounds more natural to me

Ah there is a misunderstanding here. `shutdown` is not allowed if there is a splice awaiting confirmation which can happen outside of `stfu`.

Perhaps this can be made clearer by specifying a "pending splice" instead of ongoing. I will change that.

The `shutdown` response requirement implies that the node sends `commitment_signed` to commit any outstanding changes before replying; however, it could theoretically reconnect instead, which would simply erase all outstanding uncommitted changes.

`shutdown` requires that there be no splice in progress, but if there
is already a splice in progress, it might require another splice to
"unstick" it (if the first splice was invalid, double-spent, or simply
had too low a fee), so in this case initiating another splice is legal
even after sending a shutdown.

### Closing Negotiation: `closing_signed`

> Once shutdown is complete and the channel is empty of HTLCs, the final

Reviewer: ```suggestion
Once shutdown is complete, all pending splices are confirmed, and the channel is empty of HTLCs, the final
```

   * [`signature`:`signature`]
   * [`u16`:`num_htlcs`]
   * [`num_htlcs*signature`:`htlc_signature`]
   * [`commitment_signed_tlvs`:`tlvs`]

1. `tlv_stream`: `commitment_signed_tlvs`
2. types:
    1. type: 0 (`splice_commitsigs`)
    2. data:
        * [`...*commitsigs`:`sigs`]

1. subtype: `commitsigs`
2. data:
   * [`signature`:`commit_signature`]
   * [`u16`:`num_htlcs`]

>    * [`num_htlcs*signature`:`htlc_signature`]

Reviewer 4: We can't do that unfortunately, we won't have enough space in the `commit_sig` message since it's restricted to 65kB...

We can use several `commit_sig` messages though, one for each version of the commit tx with one tlv field to indicate which splice these signatures are valid for (probably identified with just the corresponding funding txid or using an implicit ordering by feerate). We can mandate that if N splices are ongoing, senders must send (N+1) `commit_sig` messages in a row and the receiver treats them together atomically.

We can do the same for `revoke_and_ack`, or use a single message since in that case there should be enough space left in the tlv stream.

Author: Happy with multiple messages, sure.

Reviewer: Do we have an example of how splitting over multiple messages works elsewhere in the spec or would we have to design something specific for commit_sig to fit into 65kB chunks?

#### Requirements

A sending node:
  - MUST NOT send a `commitment_signed` message that does not include any

> updates, or add or remove splices.

Reviewer: Should we update to this to just "or add splices"? I don't think you can realistically remove a pending splice.

> updates, or add or remove splices.

Reviewer: ```suggestion
updates or added splices.
```

    to the ordering of the commitment transaction (see [BOLT #3](03-transactions.md#transaction-input-and-output-ordering)).
  - if it has not recently received a message from the remote node:
      - SHOULD use `ping` and await the reply `pong` before sending `commitment_signed`.

>   - MUST send a `commitsigs` for each splice in progress, in increasing feerate order.

Reviewer 5: what does `in progress` mean here? a splice that is either waiting for negotiation to complete or is awaiting conf?

Reviewer: It means splices waiting for confirmation.

Reviewer: Going to make that more clear 👍

      - MUST fail the channel.
  - if any `htlc_signature` is not valid for the corresponding HTLC transaction OR non-compliant with LOW-S-standard rule <sup>[LOWS](https://github.com/bitcoin/bitcoin/pull/6769)</sup>:
    - MUST fail the channel.

>   - if there is not exactly one `commitsigs` for each splice in progress:

Reviewer 12: As far as I can tell, `commitsigs` (quoted) isn't actually defined in this proposal? To what is it referring?

Reviewer: It is referring to `commitment_signed`'s `signature` field

Reviewer 4: That should just be:

```suggestion
  - if there is not exactly one `commit_sig` message for each splice in progress:
```

This is much clearer?

    - MUST reply with `stfu` once it can do so.

Upon disconnection:
  - the channel is no longer considered quiescent.

### Rationale

The normal use would be to cease sending updates, then wait for all
the current updates to be acknowledged by both peers, then start
quiescence.  For some protocols, choosing the initiator matters,
so this flag is sent.

If both sides send `stfu` simultaneously, they will both set
`initiator` to `1`, in which case the "initiator" is arbitrarily
considered to be the channel funder (the sender of `open_channel`).
The quiescence effect is exactly the same as if one had replied to the
other.

## Splicing

Splicing is the term given for replacing the funding transaction with
a new one.  For simplicity, splicing takes place once a channel is
[quiescent](#channel-quiescence).

Operation returns to normal once negotiation is done (while waiting
for the splice transaction(s) to confirm).

The splice is finally terminated when both sides send
`splice_locked` to indicate that one of the splice transactions
reached acceptable depth.

### The `splice` Message

1. type: 80 (`splice`)
2. data:
   * [`channel_id`:`channel_id`]

>    * [`chain_hash`:`chain_hash`]

Reviewer: @rustyrussell "Yes that is redundant now"

   * [`s64`:`relative_amount`]
   * [`u32`:`funding_feerate_perkw`]
   * [`u32`:`locktime`]
   * [`point`:`funding_pubkey`]

1. type: 81 (`splice_ack`)
2. data:
   * [`channel_id`:`channel_id`]
   * [`chain_hash`:`chain_hash`]
   * [`s64`:`relative_amount`]
   * [`point`:`funding_pubkey`]

`funding_satoshis` is the amount the sender is putting into the
post-splice channel. It includes their old funded channel amount
plus any satoshis they intend to add, or, less any satoshis they
intend to remove.

#### Requirements

The sender of `splice`:
- MUST have successfully initiated quiescence.
- MUST NOT send `splice` before sending and receiving `channel_ready`.
- MUST NOT send `splice` while a splice is being negotiated.

> - If any splice is in progress:

Reviewer 4: Maybe that's worth re-wording to explain that `in progress` here means that `splice_locked` hasn't been exchanged yet?

>   - MUST NOT send a splice message with `funding_feerate_perkw` which is less than 1.25 the previous `funding_feerate_perkw` (rounded down).

Reviewer 4: I think we shouldn't even send `splice` at all in that case, we should send `tx_init_rbf`, because we should double-spend the current (not yet `splice_locked`) splice, not spend from it.

Same in the other sections, when a pending splice hasn't been locked yet, we must use `tx_init_rbf` to propose another one.

The receiver of `splice`:
- SHOULD fail the connection if there is an pending splice, and the `funding_feerate_perkw` is not at least 1.25 the previous `funding_feerate_perkw` (rounded down).
- MUST respond with `splice_ack` containing its own `funding_pubkey`.
- MUST begin splice negotiation.

The receiver of `splice_ack`:
- MUST begin splice negotiation.

#### Rationale

There is no harm in agreeing to a splice with a high feerate
(presumably the recipient will not contribute to the splice which they
consider to be overpaying).

## Splice Negotiation

> The splice negotiation is very similar to the `tx_init_rbf` negotiation:

Reviewer 4: I think it's confusing the compare it to `tx_init_rbf`...we should just say that it uses the `interactive-tx` protocol to create a new funding transaction spending the latest funding transaction?

both sides alternate sending `tx_add_input` and `tx_add_output` until
they both send consecutive `tx_complete`.

### Requirements

(Note that many of these messages have their own, additional
requirements detailed elsewhere).

The initiator:

> - MUST `tx_add_input` an input which spends the current funding transaction output.

Reviewer 4: This isn't enough: we must also add special TLVs to `tx_add_input` to specifically mark that input as being the current channel outpoint, for which we **don't need to transmit `prevtx`** (otherwise any channel that has a funding transaction larger than 65kB could not use splice at all). In eclair, we created this TLV:

```txt
1. type: xx (`shared_channel_input`)
2. data:
   * [32*byte:`funding_txid`]
```

Both sides know the txid of the current funding transaction, this is sufficient instead of sending `prevtx`.

- MUST `tx_add_output` a zero-value output which pays to the two funding keys using the higher of the two `generation` fields.
- MUST pay for the common fields.

Each node:

> - MUST pay for their own added inputs and outputs.

Reviewer 4: This can be removed, it's part of the interactive-tx requirements already.

Upon receipt of consecutive `tx_complete`s, each node:
- MUST fail negotiation if there is not exactly one input spending the current funding transaction.

> - MUST fail negotiation if there is not exactly one output with zero value paying to the two funding keys (a.k.a. the new channel funding output)

Reviewer 4: Remove as well.

> - MUST calculate the channel capacity for each side:

Reviewer 4: That's not how it works anymore since we started using a signed funding contribution (that is a delta of the what each peer is adding or removing), that section should be re-worked.

  - Start with the previous balance
  - Add that side's new inputs (excluding the one spending the current funding transaction)
  - Subtracting each sides new outputs (except the zero-value one paying to the funding keys)
  - Subtract the total fee that side is paying for the splice transaction.
- MUST replace the zero-value funding output amount with the total channel capacity.
- MUST calculate the channel balance for each side:
  - Subtract any outstanding HTLCs offered by that side.
- If either side has added an output other than the new channel funding output:

>   - MUST fail the negotiation if the balance for that side is less than 1% of the total channel capacity.

Reviewer 4: I don't think this is accurate either: what we want to capture here is that nodes cannot splice-out below their reserve, right? This should be based on the `relative_amount` in `splice` / `splice_ack`.

When a node has changes for the remote commitment, it can apply them,
sign the resulting transaction (as defined in [BOLT #3](03-transactions.md)), and send a
`commitment_signed` message bundle. First it will send `commitment_signed`
for the active channel, then it will send another `commitment_signed` message
for each splice candidate awaiting confirmation.

1. type: 132 (`commitment_signed`)
2. data:
   * [`channel_id`:`channel_id`]
   * [`signature`:`signature`]
   * [`u16`:`num_htlcs`]
   * [`num_htlcs*signature`:`htlc_signature`]
   * [`commitment_signed_tlvs`:`tlvs`]

1. `tlv_stream`: `commitment_signed_tlvs`
2. types:
    1. type: 0 (`splice_info`)
    2. data:

>         * [`channel_id`:`splice_channel_id`]

Reviewer 4: As previously discussed, I think we should rather use a `batch_size` here, because otherwise you cannot properly deal with concurrency issues between `splice_locked` and `commit_sig` and may deadlock.

We need to specify in which order `commit_sig`s are sent in that "batch": eclair currently sends them ordered by sending the `commit_sig` matching the latest splice first, and the older ones afterwards.
